src/permission-set.js

```diff
@@ -16,523 +16,809 @@
+   * @private
+   * @param auth {Authorization}
+   * @return {PermissionSet} Returns self (chainable)
    */
-  this.rdf = options.rdf
+  addAuthorization (auth) {
+    var hashFragment = auth.hashFragment()
+    if (hashFragment in this.authorizations) {
+      // An authorization for this agent and resource combination already exists
+      // Merge the incoming access modes with its existing ones
+      this.authorizations[ hashFragment ].mergeWith(auth)
+    } else {
+      this.authorizations[ hashFragment ] = auth
+    }
+    if (!auth.virtual && auth.allowsControl()) {
+      // If acl:Control is involved, ensure implicit rules for the .acl resource
+      this.addControlPermissionsFor(auth)
+    }
+    // Create the appropriate indexes
+    this.addToAgentIndex(auth.webId(), auth.accessType, auth.resourceUrl, auth)
+    if (auth.isPublic()) {
+      this.addToPublicIndex(auth.resourceUrl, auth.accessType, auth)
+    }
+    return this
+  }
+
   /**
-   * Whether this permission set is for a 'container' or a 'resource'.
-   * Determines whether or not the inherit/'acl:default' attribute is set on
-   * all its Authorizations.
-   * @property resourceType
-   * @type {String}
+   * Creates an Authorization with the given parameters, and passes it on to
+   * `addAuthorization()` to be added to this PermissionSet.
+   * Essentially a convenience factory method.
+   * @method addAuthorizationFor
+   * @private
+   * @param resourceUrl {String}
+   * @param inherit {Boolean}
+   * @param agent {String|Quad} Agent URL (or `acl:agent` RDF triple).
+   * @param accessModes {String} 'READ'/'WRITE' etc.
+   * @param [origins=[]] {Array<String>} List of origins that are allowed access
+   * @param [mailTos=[]] {Array<String>}
+   * @return {Authorization}
    */
-  this.resourceType = isContainer ? CONTAINER : RESOURCE
+  addAuthorizationFor (resourceUrl, inherit, agent, accessModes, origins = [],
+                       mailTos = []) {
+    let auth = new Authorization(resourceUrl, inherit)
+    auth.setAgent(agent)
+    auth.addMode(accessModes)
+    auth.addOrigin(origins)
+    mailTos.forEach(mailTo => {
+      auth.addMailTo(mailTo)
+    })
+    this.addAuthorization(auth)
+    return auth
+  }
+
   /**
-   * The URL of the resource for which these permissions apply.
-   * @property resourceUrl
-   * @type {String}
+   * Adds a virtual (will not be serialized to RDF) authorization giving
+   * Read/Write/Control access to the corresponding ACL resource if acl:Control
+   * is encountered in the actual source ACL.
+   * @method addControlPermissionsFor
+   * @private
+   * @param auth {Authorization} Authorization containing an acl:Control access
+   *   mode.
    */
-  this.resourceUrl = resourceUrl
+  addControlPermissionsFor (auth) {
+    let impliedAuth = auth.clone()
+    impliedAuth.resourceUrl = this.aclUrlFor(auth.resourceUrl)
+    impliedAuth.virtual = true
+    impliedAuth.addMode(Authorization.ALL_MODES)
```

Comment:
Actually, no. If you look at the implementation of `addMode()`, it's not going to hit the clone issue.